File: src/peated/models/bottle.py
from sqlalchemy import (
    BigInteger,
    Column,
    DateTime,
    Enum,
    ForeignKeyConstraint,
    Index,
    PrimaryKeyConstraint,
    SmallInteger,
    String,
    text,
)

from peated.db.base import Base


class Bottle(Base):
    __tablename__ = "bottle"
    __table_args__ = (
        ForeignKeyConstraint(["bottler_id"], ["entity.id"], name="bottle_bottler_id_entity_id_fk"),
        ForeignKeyConstraint(["brand_id"], ["entity.id"], name="bottle_brand_id_entity_id_fk"),
        ForeignKeyConstraint(["created_by_id"], ["user.id"], name="bottle_created_by_id_user_id_fk"),
        PrimaryKeyConstraint("id", name="bottle_pkey"),
        Index("bottle_brand_unq", "name", "brand_id", unique=True),
        Index("bottle_name_unq", "full_name", unique=True),
    )

    id = Column(BigInteger)
    name = Column(String(255), nullable=False)
    brand_id = Column(BigInteger, nullable=False)
    created_at = Column(DateTime, nullable=False, server_default=text("now()"))
    created_by_id = Column(BigInteger, nullable=False)
    total_tastings = Column(BigInteger, nullable=False, server_default=text("0"))
    full_name = Column(String(255), nullable=False)
    category = Column(
        Enum(
            "blend",
            "bourbon",
            "rye",
            "single_grain",
            "single_malt",
            "spirit",
            name="category",
        )
    )
    stated_age = Column(SmallInteger)
    bottler_id = Column(BigInteger)

File: src/server/database/model/bottle.go
package model

import (
	"time"

	"github.com/lib/pq"
	"gorm.io/datatypes"
)

const (
	CategoryBlend          string = "blend"
	CategoryBourbon        string = "bourbon"
	CategoryRye            string = "rye"
	CategorySingleGrain    string = "single_grain"
	CategorySingleMalt     string = "single_malt"
	CategorySinglePotStill string = "single_pot_still"
	CategorySpirit         string = "spirit"
)

var CategoryNames = []string{
	CategoryBlend,
	CategoryBourbon,
	CategoryRye,
	CategorySingleGrain,
	CategorySingleMalt,
	CategorySinglePotStill,
	CategorySpirit,
}

type Bottle struct {
	ID        uint64  `gorm:"primaryKey" json:"id"`
	FullName  string  `json:"full_name" gorm:"not null"`
	Name      string  `json:"name" gorm:"not null"`
	Category  *string `json:"category"`
	BrandID   uint64  `json:"brand_id" gorm:"not null"`
	BottlerID *uint64 `json:"bottler_id"`
	StatedAge *uint   `json:"stated_age"`

	Description   *string         `json:"description"`
	TastingNotes  *datatypes.JSON `json:"tasting_notes"`
	SuggestedTags *pq.StringArray `json:"suggested_tags" gorm:"type:string[];default:[];not null"`

	AvgRating     *float64 `json:"avg_rating"`
	TotalTastings uint     `json:"total_tastings" gorm:"default:0;not null"`

	CreatedAt   time.Time `json:"created_at"`
	CreatedByID uint64    `json:"created_by_id"`

	Brand      Entity
	Bottler    Entity
	Distillers []Entity `gorm:"many2many:bottle_distiller"`
	CreatedBy  User
}

func (Bottle) TableName() string {
	return "bottle"
}

type Bottles []*Bottle
